File: python/leetcode/141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycleHashMap(self, head: Optional[ListNode]) -> bool:

        hash = {}
        slow = head

        while slow:
            if not slow.next or not slow.next.next:
                return False

            if slow in hash:
                return True
            else:
                hash[slow] = 1
            slow = slow.next

class Solution:
    def hasCycleTwoPointers(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return False
        slow, fast = head, head.next
        
        while slow and fast:
            if not fast:
                return False
            elif not fast.next or not fast.next.next:
                return False
            
            if fast == slow:
                return True
            
            slow = slow.next
            fast = fast.next.next

File: c#/leetcode/141-linked-list-cycle.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution
{
  public bool HasCycle(ListNode head)
  {
    HashSet<ListNode> hash = new HashSet<ListNode>();

    ListNode current = head;

    while (current != null)
    {
      if (hash.Contains(current))
      {
        return true;
      }
      hash.Add(current);
      current = current.next;
    }
    return false;
  }

  public bool HasCycleDict(ListNode head)
  {
    var dict = new Dictionary<ListNode, int>();

    ListNode current = head;

    while (current != null)
    {
      if (dict.ContainsKey(current))
      {
        return true;
      }
      else
      {
        dict.Add(current, 1);
      }
      current = current.next;
    }
    return false;
  }

  public bool HasCycleTwoPointers(ListNode head)
  {
    if (head == null || head.next == null)
    {
      return false;
    }

    ListNode slow = head;
    ListNode fast = head.next;

    while (slow != null && fast != null)
    {
      if (fast.next == null || fast.next.next == null)
      {
        return false;
      }

      if (fast == slow)
      {
        return true;
      }

      slow = slow.next;
      fast = fast.next.next;
    }
    return false;
  }
}